File: src/main/java/ninja/leaping/permissionsex/util/command/Commander.java
/**
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ninja.leaping.permissionsex.util.command;

import com.google.common.base.Optional;
import ninja.leaping.permissionsex.util.Translatable;

import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * Created by zml on 20.03.15.
 */
public interface Commander<TextType> {
    public String getName();
    public Locale getLocale();
    public Optional<Map.Entry<String, String>> getSubjectIdentifier();
    public boolean hasPermission(String permission);
    public Set<Map.Entry<String, String>> getActiveContexts();
    public MessageFormatter<TextType> fmt();
    public void msg(Translatable text);
    public void debug(Translatable text);
    public void error(Translatable text);
    public void msg(TextType text);
    public void debug(TextType text);
    public void error(TextType text);
}

File: permissionsex-bukkit/src/main/kotlin/ca/stellardrift/permissionsex/bukkit/Commander.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.bukkit

import ca.stellardrift.permissionsex.PermissionsEx
import ca.stellardrift.permissionsex.commands.commander.MessageFormatter
import ca.stellardrift.permissionsex.commands.commander.Commander
import ca.stellardrift.permissionsex.util.*
import com.google.common.collect.Maps
import net.kyori.text.BuildableComponent
import net.kyori.text.Component
import net.kyori.text.ComponentBuilder
import net.kyori.text.TextComponent
import net.kyori.text.TextComponent.newline
import net.kyori.text.TextComponent.space
import net.kyori.text.adapter.bukkit.TextAdapter
import net.kyori.text.event.ClickEvent
import net.kyori.text.event.HoverEvent
import net.kyori.text.format.TextColor
import net.kyori.text.format.TextDecoration
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import java.util.Locale
import java.util.Optional

fun Iterable<CommandSender>.sendMessage(text: Component) = TextAdapter.sendComponent(this, text)
fun Iterable<CommandSender>.sendActionBar(text: Component) = TextAdapter.sendActionBar(this, text)

fun CommandSender.sendMessage(text: Component) = TextAdapter.sendComponent(this, text)
fun CommandSender.sendActionBar(text: Component) = TextAdapter.sendActionBar(this, text)

/**
 * Take a locale string provided from a minecraft client and attempt to parse it as a locale.
 * These are not strictly compliant with the iso standard, so we try to make things a bit more normalized.
 *
 * @param mcLocaleString The locale string, in the format provided by the Minecraft client
 * @return A Locale object matching the provided locale string
 */
fun String.toLocale(): Locale {
    val parts = this.split("_", limit = 3)
    return when (parts.size) {
        0 -> Locale.getDefault();
        1 -> Locale(parts[0]);
        2 -> Locale(parts[0], parts[1]);
        3 -> Locale(parts[0], parts[1], parts[2]);
        else -> throw IllegalArgumentException("Provided locale '$this' was not in a valid format!");
    }
}

class BukkitMessageFormatter(private val cmd: BukkitCommander): MessageFormatter(cmd.pex.manager) {

    override val Map.Entry<String, String>.friendlyName: String?
        get() = cmd.pex.manager.getSubjects(key).typeInfo.getAssociatedObject(value).castMap<CommandSender, String> { name }

}

/**
 * An abstraction over the Sponge CommandSource that handles PEX-specific message formatting and localization
 */
class BukkitCommander internal constructor(
    internal val pex: PermissionsExPlugin,
    private val commandSource: CommandSender
) : Commander {
    override val formatter: BukkitMessageFormatter = BukkitMessageFormatter(this)
    override val name: String
        get() = commandSource.name

    override fun hasPermission(permission: String): Boolean = commandSource.hasPermission(permission)

    override val locale: Locale
        get() = if (commandSource is Player) commandSource.locale.toLocale() else Locale.getDefault()

    override val subjectIdentifier: SubjectIdentifier?
        get() = if (commandSource is Player) {
                subjectIdentifier(
                    PermissionsEx.SUBJECTS_USER,
                    commandSource.uniqueId.toString()
                )
        } else null

    override fun msg(text: Component) {
        commandSource.sendMessage(PEXComponentRenderer.render(text coloredIfNecessary TextColor.DARK_AQUA, locale))
    }
}
